cilium/CFP-33716-distributed-packet-tracing.md

```diff
@@ -0,0 +1,586 @@
+
+Additionally, within Cilium, these flows will generate datapath debugging events regardless of the global debug-verbose setting. This will improve debuggability of production systems.
+
+## Motivation
+There are currently limitations for gathering and correlating flow events:
+- NAT transformations and DSR can all change the 4-tuple, so that a single Hubble or tcpdump filter cannot be used to trace the packet along the entire path – even when using a system such as Hubble Relay.
+- Enabling verbose datapath debugging is not possible in production environments due to traffic volume which leads to dropping flows.
+
+## Goals
+- For tagged packets, add the ID to emitted flow events.
+- For tagged packets, emit datapath debugging events regardless of the debug-verbose setting.
+- Support both IPv4 and IPv6.
+- Add hubble filter for trace ID.
+- Allow users to define which flow IDs they want to monitor for events.
+
+## Non-goals
+- Automatically apply IP option to packets (see Future Milestones).
+
+## User Stories
+- As a platform administrator managing multiple clusters, users are reporting connection timeouts reaching destinations in remote clusters. Platform admin generates traffic with trace ID IP option along the problematic path to identify all flow events associated with that packet at both the source and destination.
+
+![User story](./images/CFP-33716-user-story.png)
+
+- As a production cluster administrator, particular flows from a client are experiencing connection timeouts (e.g. [issue #32472](https://github.com/cilium/cilium/issues/32472). Enabling datapath debugging for all flows leads to event loss and low SNR. To extract dataplane information to root cause the issue, generate traffic with trace ID IP option to emit datapath debug information specific to this flow.
+
+## Proposal
+
+### Overview
+Offer support for packets containing a user-defined IP option field. For tagged incoming flows, handle events for this flow specially; emit datapath debugging events for these packets, include the ID in the event message, and implement hubble filtering by the ID.
+
+### Section 1: Packet Tagging 
+### Packet Tagging Responsibility
+
+This feature does not implement the packet tagging (see Deferred Milestone 1: Tag Packets). Instead, it is the client's responsibility to tag packets via IP option embedding. Therefore, the key assumption for this feature is that tagged packets arrive at ingress points with an `ip_trace_id` already embedded in the IP options.
+
+#### Reasons for Not Implementing Automatic Packet Tagging:
+
+1. **Avoid MTU Concerns:**
+   - Automatically tagging packets can cause them to exceed the MTU, leading to fragmentation and performance issues.
```

Comment:
For an item like this, it would be good to give a reason of {mitigation || why this is not a problem}, etc. so the reader can understand the thought process here.